Analytics reporting support

delroth opened the pull request:
<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/3905)
<!-- Reviewable:end -->

MaJoRoesch commented:
Review status: 0 of 18 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWBYGGcbCAK9BqrOBY:-KKWBYGGcbCAK9BqrOBZ:-928694738) (no related file):*
![scary](http://i.imgur.com/lTl3pAZ.png)

"Analytics" is a bit of a scary word, since it is usually applied to ad collection. We had a bikeshedding session about this last night for the article, and "Statistic Reporting" is what we settled on as being a gentle but accurate term. What do you think delroth?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905)*
<!-- Sent from Reviewable.io -->

delroth commented:
Review status: 0 of 302 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWBYGGcbCAK9BqrOBY:-KKWJ123ur5a3KajcIBn:1461777776) (no related file):*
<details><summary><i>Previously, MaJoRoesch (MaJoR) wrote…</i></summary>
> ![scary](http://i.imgur.com/lTl3pAZ.png)
> 
> "Analytics" is a bit of a scary word, since it is usually applied to ad collection. We had a bikeshedding session about this last night for the article, and "Statistic Reporting" is what we settled on as being a gentle but accurate term. What do you think delroth?
</details>
It's now "Usage statistics reporting".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905)*
<!-- Sent from Reviewable.io -->

shellashock commented:
How exactly would this system send data to the server? Would it send a small packet of information as soon as Dolphin/ a game is launched or does it wait until Dolphin has been closed before sending all the accumulated data? I would think probably the former, but just curious as I can't test until Windows support is added.

lioncash commented:
Review status: 0 of 302 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[Source/Core/Common/Analytics.cpp, line 192 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWkeU2oP5f7m7icVxz:-KKWkeU3vTwNS3BfvuSh:319738030) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Common/Analytics.cpp#L192)):*
> ```cpp
> HttpAnalyticsBackend::HttpAnalyticsBackend(const std::string& endpoint)
> {
> 	CURL* curl = curl_easy_init();
> ```

Just a suggestion, but this can technically use a unique_ptr with a custom deleter to do automatic tear-down (I've put both methods of doing so [here](https://ideone.com/zpScxm), just adapt pointer types where necessary instead of `void*`).  The benefit of the concrete struct-based deleter is you essentially make a full general-purpose wrapper for CURL* pointers that can be used anywhere.

---

*[Source/Core/Common/Analytics.h, line 103 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWqzeDfpsvt2SVrd0X:-KKWqzeDfpsvt2SVrd0Y:-1685377231) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Common/Analytics.h#L103)):*
> ```objc
> 	}
> 
> 	std::string Get() const {
> ```

Braces, same for `Consume` below.

---

*[Source/Core/Common/Analytics.h, line 138 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWrBLwFtM2KZQtMiLf:-KKWrBLxjZ0wGxKjcJ3g:598515237) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Common/Analytics.h#L138)):*
> ```objc
> 	void SetBackend(std::unique_ptr<AnalyticsReportingBackend> backend)
> 	{
> 		m_backend.reset(backend.release());
> ```

```cpp
m_backend = std::move(backend);
```

---

*[Source/Core/Core/Analytics.cpp, line 24 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWvMU_A0uicX2KNHfP:-KKWvMUadDQDdILPe8JU:1105966973) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Core/Analytics.cpp#L24)):*
> ```cpp
> namespace
> {
> const char* ANALYTICS_ENDPOINT = "https://analytics.dolphin-emu.org/report";
> ```

This can actually be `constexpr const char*`, which will just store the string data rather than the string data and a pointer (e.g. [example](https://godbolt.org/g/CaHBOl)). This is super inconsequential, so feel free to ignore it.

---

*[Source/Core/Core/Analytics.cpp, line 41 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWy1t_cfLh90t-QrOE:-KKWy1t_cfLh90t-QrOF:-1934559106) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Core/Analytics.cpp#L41)):*
> ```cpp
> 	if (!s_instance)
> 	{
> 		s_instance.reset(new DolphinAnalytics());
> ```

This can be:

```cpp
s_instance = std::make_shared<DolphinAnalytics>();
```

---

*[Source/Core/Core/Analytics.cpp, line 54 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWwfMSG3LoD0u7cA8K:-KKWwfMTi93qvnFzeLEA:1558134003) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Core/Analytics.cpp#L54)):*
> ```cpp
> 	if (SConfig::GetInstance().m_analytics_enabled)
> 	{
> 		new_backend.reset(new Common::HttpAnalyticsBackend(ANALYTICS_ENDPOINT));
> ```

This can be:

```cpp
new_backend = std::make_unique<Common::HttpAnalyticsBackend>(ANALYTICS_ENDPOINT);
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905)*
<!-- Sent from Reviewable.io -->

degasus commented on Source/Core/Common/StringUtil.cpp:
```diff
@@ -36,6 +36,37 @@ static locale_t GetCLocale()
 
+std::string HexDump(const u8* data, size_t size)
+{
+	std::string out;
```
What do you think about pre-allocating?

degasus commented on Source/Core/Common/StringUtil.cpp:
```diff
@@ -36,6 +36,37 @@ static locale_t GetCLocale()
+std::string HexDump(const u8* data, size_t size)
+{
+	std::string out;
+	for (size_t row_start = 0; row_start < size; row_start += 16)
```
Please use a constant for "16". characters_per_line ?

degasus commented on Source/Core/Common/Analytics.h:
```diff
@@ -0,0 +1,190 @@
+	}
+
+protected:
+	static void AppendSerializedValue(std::string* report, const std::string& v);
```
template <typename T> static void AppendSerializedValue(std::string* report, const T& v);

degasus commented on Source/Core/Common/Analytics.cpp:
```diff
@@ -0,0 +1,221 @@
+	FLOAT = 4,
+};
+
+void AppendBool(std::string* out, bool v)
```
static

And also with a templated definition.

comex commented:
> "Analytics" is a bit of a scary word, since it is usually applied to ad collection. We had a bikeshedding session about this last night for the article, and "Statistic Reporting" is what we settled on as being a gentle but accurate term. What do you think delroth?

What about a name similar to the Steam Hardware Survey?  Though that's a bit different since apparently it explicitly pops up to survey you.

Sorry to nitpick, but what is the point of using some random custom binary format rather than something standard like JSON?  These are not big packets, and a simple JSON serializer is only a few lines of code.

phire commented:
Reviewed 2 of 2 files at r1, 285 of 285 files at r2, 3 of 3 files at r3, 11 of 11 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/Analytics.cpp, line 70 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXH6nMd8YUh6iQNGuD:-KKXH6nMd8YUh6iQNGuE:-1113201056) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Analytics.cpp#L70)):*
> ```cpp
> 	std::random_device rd;
> 	u64 id = (static_cast<u64>(rd()) << 32) | rd();
> 	m_unique_id = StringFromFormat("%016" PRIx64, id);
> ```

You might want to consider including a 8bit checksum (XOR the other 7 bytes together). No need to check for it now, but it might prove useful in the future.

---

*[Source/Core/Core/Analytics.cpp, line 138 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXI5-l4zn_z7R2fplk:-KKXI5-mZYL4_gDUnRLW:-749544382) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Analytics.cpp#L138)):*
> ```cpp
> 	builder.AddData("osx-ver-bugfix", osxbugfix);
> #elif defined(__linux__)
> 	builder.AddData("os-type", "linux");
> ```

Kernel and libc version?

---

*[Source/Core/Core/Analytics.cpp, line 175 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXJB1KwJyAwIhWwI6R:-KKXJB1KwJyAwIhWwI6S:-1105020435) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Analytics.cpp#L175)):*
> ```cpp
> 	builder.AddData("cfg-gfx-anisotropy", g_Config.iMaxAnisotropy);
> 	builder.AddData("cfg-gfx-realxfb", g_Config.RealXFBEnabled());
> 	builder.AddData("cfg-gfx-virtualxfb", g_Config.VirtualXFBEnabled());
> ```

Can I also request:
 * bVsync
 * bFullScreen
 * bExclusiveMode
 * iAspectRatio
 * bEFBAccessEnable
 * bEFBEmulateFormatChagnes
 * bSkipEFBCopyToRam
 * bCopyEFBScaled
 * iSafeTextureCache_ColorSamples;
 * iStereoMode
 * sPostProcessingShader

 * s_video_backend->GetName()
 * g_Config.Adapters[g_Config.iAdapter] // Selected Video Card. directx 11/12 only, make sure the vector isn't empty first.
 * g_ogl_config.{gl_vendor, gl_renderer, gl_version} // OpenGL equivalent (OGL/Render.cpp line 591)

It would also be nice to get all of bSupports*, but that only needs to be sent once per backend (per driver update)

---

*[Source/Core/Core/Core.cpp, line 549 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXWr3d0CecAwgciOGe:-KKXWr3d0CecAwgciOGf:1270750480) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Core.cpp#L549)):*
> ```cpp
> 	s_is_booting.store(false);
> 
> 	DolphinAnalytics::Instance()->ReportGameStart();
> ```

BTW, I could see merit in delaying 30 seconds after the game starts before reporting it. 

---

*[Source/Core/DolphinWX/Main.cpp, line 273 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXYHYabvaT520e1ym6:-KKXYHYabvaT520e1ym7:-576403798) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/DolphinWX/Main.cpp#L273)):*
> ```cpp
> 	{
> 		int answer = wxMessageBox(
> 				_("Dolphin occasionally collects data about the hardware, operating "
> ```

BTW "occasionally" is not quite the same as "Every single boot and game launch" 

---

*[Source/Core/DolphinWX/Main.cpp, line 274 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXXm7z1PZhYMHZiA1F:-KKXXm7z1PZhYMHZiA1G:-81561157) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/DolphinWX/Main.cpp#L274)):*
> ```cpp
> 		int answer = wxMessageBox(
> 				_("Dolphin occasionally collects data about the hardware, operating "
> 				  "system, and application configuration used to run the emulator. "
> ```

Do we want to mention "Performance characteristics and which games are using which emulated features" despite the fact we are not yet measuring them.

Or should we change it later and pop up a new permission dialog?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905)*
<!-- Sent from Reviewable.io -->

phire commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[Source/Core/DolphinWX/Main.cpp, line 273 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXYHYabvaT520e1ym6:-KKX_7lz8bBdgxfNZm00:-662166800) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/DolphinWX/Main.cpp#L273)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> BTW "occasionally" is not quite the same as "Every single boot and game launch" 
</details>
"Optionally" would fit better here, and is actually accurate.

---

*[Source/Core/DolphinWX/Main.cpp, line 278 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKX_J9S6YEzxxip-3-G:-KKXXaLKLT_KtUOof5-t:-1654959392) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/DolphinWX/Main.cpp#L278)):*
> ```cpp
> 				  "how people use Dolphin and prioritize our efforts.\n\n"
> 				  "Do you authorize Dolphin to report this information to Dolphin "
> 				  "developers?"),
> ```

Add: "You can revoke this authorization later in settings"

That will encourage a few more people who are on the fence to click yes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905)*
<!-- Sent from Reviewable.io -->

JMC47 commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/Core.cpp, line 549 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXWr3d0CecAwgciOGe:-KKXbtxW9EnyWI1O1DrQ:-690728111) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Core.cpp#L549)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> BTW, I could see merit in delaying 30 seconds after the game starts before reporting it. 
</details>
But then games I boot up and then completely facepalm on won't show up.  Also games that crash on boot won't show up either, which may be useful to know.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905)*
<!-- Sent from Reviewable.io -->

phire commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/Core.cpp, line 549 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXWr3d0CecAwgciOGe:-KKXfwQmgJuac9pV1PJX:-1725498660) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Core.cpp#L549)):*
<details><summary><i>Previously, JMC47 wrote…</i></summary>
> But then games I boot up and then completely facepalm on won't show up.  Also games that crash on boot won't show up either, which may be useful to know.
</details>
We don't have enough information here to tell that anyway, 

A 30 second delay would be useful for filtering out the accidental launches or whenever someone realises they need to select new settings first.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905)*
<!-- Sent from Reviewable.io -->

BhaaLseN commented:
Review status: all files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[CMakeLists.txt, line 794 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKY7wFgl79Y5bPJFZqZ:-KKY7wFgl79Y5bPJFZq_:992228018) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/CMakeLists.txt#L794)):*
> ```CMake
>   add_subdirectory(Externals/curl)
>   set(CURL_LIBRARIES curl)
>   include_directories(BEFORE Externals/curl/include)
> ```

Probably a non-issue, but indent is off here, tabs vs. spaces

---

*[Source/Core/Common/Analytics.cpp, line 218 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKY2gBNHElwjyK6DcF8:-KKY2gBNHElwjyK6DcF9:-839323895) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Common/Analytics.cpp#L218)):*
> ```cpp
> 	curl_easy_setopt(m_curl, CURLOPT_POSTFIELDS, report.c_str());
> 	curl_easy_setopt(m_curl, CURLOPT_POSTFIELDSIZE, report.size());
> 	curl_easy_perform(m_curl);
> ```

What happens if this fails (ie. because our cert changed/expired/has been revoked, or the server is down, or the user is not connected to the internet, something like that)? I do hope the answer is not "crash".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905)*
<!-- Sent from Reviewable.io -->

delroth commented:
OS X is not using the Externals libcurl right now so I haven't been able to test that. I'll do more testing there once I'm done with Windows.

Regarding the format, a JSON serializer would be more code than this, especially if we want to handle stuff like non-Unicode strings or 64 bit integer types. If we had already a JSON serializer I would have used that. We're not stuck to the current format anyway (as long as we keep the first byte for the wire format version).

As mentioned in another comment, the name is now "Usage statistics reporting".

---

Review status: all files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[CMakeLists.txt, line 794 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKY7wFgl79Y5bPJFZqZ:-KKYC8hvMqbyZTsLu4Dw:1758359711) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/CMakeLists.txt#L794)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary>
> Probably a non-issue, but indent is off here, tabs vs. spaces
</details>
Fixed.

---

*[Source/Core/Common/Analytics.cpp, line 33 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWih401W3HSN5fGJ6a-r4-33:-KKYCFFK61UR7lf1wiOx:1018284921) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Common/Analytics.cpp#L33)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> static
> 
> 
> 
> And also with a templated definition.
</details>
It's in an unnamed namespace already. I don't see any reason to have this templated?

---

*[Source/Core/Common/Analytics.cpp, line 192 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWkeU2oP5f7m7icVxz:-KKYCSpeVVWfFYBzj7Bt:-1420109919) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Common/Analytics.cpp#L192)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Just a suggestion, but this can technically use a unique_ptr with a custom deleter to do automatic tear-down (I've put both methods of doing so [here](https://ideone.com/zpScxm), just adapt pointer types where necessary instead of `void*`).  The benefit of the concrete struct-based deleter is you essentially make a full general-purpose wrapper for CURL* pointers that can be used anywhere.
</details>
I'm guessing in the near-ish future we'll have our own C++ wrappers on top of curl anyway. Right now this is simple enough, introducing unique_ptr with custom deleter doesn't really gain us much. I'll do it if you really care :)

---

*[Source/Core/Common/Analytics.cpp, line 218 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKY2gBNHElwjyK6DcF8:-KKYClprt7LJFaUQgJY_:1028704604) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Common/Analytics.cpp#L218)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary>
> What happens if this fails (ie. because our cert changed/expired/has been revoked, or the server is down, or the user is not connected to the internet, something like that)? I do hope the answer is not "crash".
</details>
This returns an error code which we ignore because we don't care ¯\_(ツ)_/¯

---

*[Source/Core/Common/Analytics.h, line 115 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWih401W3HSN5fGJ6b-r4-115:-KKYD4wwO2lFm1s58fCl:-1266929027) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Common/Analytics.h#L115)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> template <typename T> static void AppendSerializedValue(std::string* report, const T& v);
</details>
Why? This has separate implementations for each of the types, and having it be templated would require the implementation to be inlined in header files. This would increase our compilation times for no good reason.

---

*[Source/Core/Common/Analytics.h, line 103 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWqzeDfpsvt2SVrd0X:-KKYCruiFEaSl_lq9Ib9:1365291750) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Common/Analytics.h#L103)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Braces, same for `Consume` below.
</details>
Fixed. Have I mentioned how much I have the Allman brace style? :P

---

*[Source/Core/Common/Analytics.h, line 138 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWrBLwFtM2KZQtMiLf:-KKYDEVThKru-Yaa6q--:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Common/Analytics.h#L138)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> m_backend = std::move(backend);
> ```
</details>
Done.

---

*[Source/Core/Common/StringUtil.cpp, line 41 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKW6PGGm44S4PW9RDKf-r1-41:-KKYDM7p7ANFkIBVOcRk:961569979) ([raw file](https://github.com/dolphin-emu/dolphin/blob/956cd98803d029476f84ef4af2a911c4644bf6e4/Source/Core/Common/StringUtil.cpp#L41)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> What do you think about pre-allocating?
</details>
Meh. If HexDump becomes performance critical we have more problems.

---

*[Source/Core/Common/StringUtil.cpp, line 42 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKW6PGGm44S4PW9RDKf-r1-42:-KKYDQinkyiKKCdiZGsU:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/956cd98803d029476f84ef4af2a911c4644bf6e4/Source/Core/Common/StringUtil.cpp#L42)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> Please use a constant for "16". characters_per_line ?
</details>
Done.

---

*[Source/Core/Core/Analytics.cpp, line 70 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXH6nMd8YUh6iQNGuD:-KKYDwloVCEg0_HB0wPp:3029446) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Analytics.cpp#L70)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> You might want to consider including a 8bit checksum (XOR the other 7 bytes together). No need to check for it now, but it might prove useful in the future.
</details>
Why? It reduces entropy and I can't think of a purpose for it.

---

*[Source/Core/Core/Analytics.cpp, line 138 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXI5-l4zn_z7R2fplk:-KKYE0SlKvsXOs_FOUWW:520737146) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Analytics.cpp#L138)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> Kernel and libc version?
</details>
Do we expect that to ever be useful? It's more pain to collect than what it will gain us, I think.

---

*[Source/Core/Core/Analytics.cpp, line 175 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXJB1KwJyAwIhWwI6R:-KKYFUECzzODFNxEvBaX:-1427121347) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Analytics.cpp#L175)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> Can I also request:
>  * bVsync
>  * bFullScreen
>  * bExclusiveMode
>  * iAspectRatio
>  * bEFBAccessEnable
>  * bEFBEmulateFormatChagnes
>  * bSkipEFBCopyToRam
>  * bCopyEFBScaled
>  * iSafeTextureCache_ColorSamples;
>  * iStereoMode
>  * sPostProcessingShader
> 
>  * s_video_backend->GetName()
>  * g_Config.Adapters[g_Config.iAdapter] // Selected Video Card. directx 11/12 only, make sure the vector isn't empty first.
>  * g_ogl_config.{gl_vendor, gl_renderer, gl_version} // OpenGL equivalent (OGL/Render.cpp line 591)
> 
> It would also be nice to get all of bSupports*, but that only needs to be sent once per backend (per driver update)
> 
</details>
Done except for:

* sPostProcessingShader: it's a filename, I consider that user data.
* s_video_backend->GetName(): if I'm not mistaken, should be the same as s_video_backend->GetName()?
* g_ogl_config fields: No way I'll make Core depend on VideoBackends, I feel dirty enough making it depend on VideoCommon :) If you want to send a followup for that, feel free.

---

*[Source/Core/Core/Analytics.cpp, line 24 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWvMU_A0uicX2KNHfP:-KKYDagqTGV1FOTTtfJn:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Core/Analytics.cpp#L24)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This can actually be `constexpr const char*`, which will just store the string data rather than the string data and a pointer (e.g. [example](https://godbolt.org/g/CaHBOl)). This is super inconsequential, so feel free to ignore it.
</details>
Done.

---

*[Source/Core/Core/Analytics.cpp, line 41 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWy1t_cfLh90t-QrOE:-KKYDtKiF8ukj20YCwsH:-1716417977) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Core/Analytics.cpp#L41)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This can be:
> 
> ```cpp
> s_instance = std::make_shared<DolphinAnalytics>();
> ```
</details>
Unfortunately not possible since the constructor is private.

---

*[Source/Core/Core/Analytics.cpp, line 54 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWwfMSG3LoD0u7cA8K:-KKYDu8uvauh8fXfzz0f:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Core/Analytics.cpp#L54)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This can be:
> 
> ```cpp
> new_backend = std::make_unique<Common::HttpAnalyticsBackend>(ANALYTICS_ENDPOINT);
> ```
</details>
Done.

---

*[Source/Core/Core/Core.cpp, line 549 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXWr3d0CecAwgciOGe:-KKYH6UH_i3YJEKZIxKi:-1476617029) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Core.cpp#L549)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> We don't have enough information here to tell that anyway, 
> 
> A 30 second delay would be useful for filtering out the accidental launches or whenever someone realises they need to select new settings first.
</details>
Adding a 30 second delay would make things a bit more complicated. I'd rather keep the integration simple for now. Worst case if we really feel the need to, accidental launches can be filtered out on the server side (say, if two game start events in less than 1min, remove the oldest once).

---

*[Source/Core/DolphinWX/Main.cpp, line 273 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXYHYabvaT520e1ym6:-KKYHuUJggHR4-Nyr3dD:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/DolphinWX/Main.cpp#L273)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> "Optionally" would fit better here, and is actually accurate.
</details>
Done.

---

*[Source/Core/DolphinWX/Main.cpp, line 274 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXXm7z1PZhYMHZiA1F:-KKYHtjs8EipZ7h088cr:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/DolphinWX/Main.cpp#L274)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> Do we want to mention "Performance characteristics and which games are using which emulated features" despite the fact we are not yet measuring them.
> 
> Or should we change it later and pop up a new permission dialog?
</details>
Done.

---

*[Source/Core/DolphinWX/Main.cpp, line 278 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKX_J9S6YEzxxip-3-G:-KKYHszPWYjK-9KxJ67B:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/DolphinWX/Main.cpp#L278)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> Add: "You can revoke this authorization later in settings"
> 
> That will encourage a few more people who are on the fence to click yes.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-:-KKYBIcyfBCw8jb3T0IB:1919672154)*
<!-- Sent from Reviewable.io -->

phire commented:
Review status: 1 of 302 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/Analytics.cpp, line 70 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXH6nMd8YUh6iQNGuD:-KKYNaM13WvaYSHXIDTl:-2146581096) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Analytics.cpp#L70)):*
<details><summary><i>Previously, delroth (Pierre Bourdon) wrote…</i></summary>
> Why? It reduces entropy and I can't think of a purpose for it.
</details>
I don't really think we will miss the 8 bits of entropy (though 56bits is cutting things a little fine), but I think being able to identify IDs that were generated outside of dolphin by some kind of naive DDoS attack (which doesn't look at our source code) might be worth it.

---

*[Source/Core/Core/Analytics.cpp, line 138 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXI5-l4zn_z7R2fplk:-KKYOKBVn0jEyHMfPmlK:-252521191) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Analytics.cpp#L138)):*
<details><summary><i>Previously, delroth (Pierre Bourdon) wrote…</i></summary>
> Do we expect that to ever be useful? It's more pain to collect than what it will gain us, I think.
</details>
libc version is directly correlated to gcc version, which gives us an idea about how old the user's distro is and what compilers we need to support. Kernel version would also indicate age, but is less useful. 

The main question we want to ask here is "what compilers do we need to support"

---

*[Source/Core/Core/Analytics.cpp, line 175 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKXJB1KwJyAwIhWwI6R:-KKYQBUxv3nwQ-ucj5pm:-1066772948) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c31d2df63b77c69c9beb9eda769e71b7cf1c00a8/Source/Core/Core/Analytics.cpp#L175)):*
<details><summary><i>Previously, delroth (Pierre Bourdon) wrote…</i></summary>
> Done except for:
> 
> * sPostProcessingShader: it's a filename, I consider that user data.
> * s_video_backend->GetName(): if I'm not mistaken, should be the same as s_video_backend->GetName()?
> * g_ogl_config fields: No way I'll make Core depend on VideoBackends, I feel dirty enough making it depend on VideoCommon :) If you want to send a followup for that, feel free.
</details>
Agreed on all points.

Perhaps we should have the opengl backend report this information back to video common (like direct x currently does)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905)*
<!-- Sent from Reviewable.io -->

mbc07 commented:
What about translations?

Helios747 commented:
You're not going to like the answer to this. None of us did.

JMC47 commented:
Google Translate.

RisingFog commented:
@JMC47 *yak*

mbc07 commented:
Well, I know we are in string freeze but maybe we can push this to Transifex with a deadline of a few hours? From what I can tell there are just 3 or 4 new sentences...

delroth commented:
@mbc07 AFAIK, that's the plan. We have some time before other stuff (articles, video, etc.) are ready for the release anyway, which should let time for translators to go through this.

lioncash commented:
Review status: 1 of 313 files reviewed at latest revision, 14 unresolved discussions.

---

*[Source/Core/Common/Analytics.cpp, line 192 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWkeU2oP5f7m7icVxz:-KKZPJ8QSkAT8jGAcuXE:236709316) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Common/Analytics.cpp#L192)):*
<details><summary><i>Previously, delroth (Pierre Bourdon) wrote…</i></summary>
> I'm guessing in the near-ish future we'll have our own C++ wrappers on top of curl anyway. Right now this is simple enough, introducing unique_ptr with custom deleter doesn't really gain us much. I'll do it if you really care :)
</details>
Yeah, I can just make a wrapper header for curl that goes in Common after 5.0 is released or something, (I had meant to mark my initial comment as satisfied/non-blocking)

---

*[Source/Core/Core/Analytics.cpp, line 41 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905#-KKWy1t_cfLh90t-QrOE:-KKZP5CnwtBKBNdwUMUj:430291858) ([raw file](https://github.com/dolphin-emu/dolphin/blob/b616237f0b384b20ece21b9a9b6b249d48db5ee9/Source/Core/Core/Analytics.cpp#L41)):*
<details><summary><i>Previously, delroth (Pierre Bourdon) wrote…</i></summary>
> Unfortunately not possible since the constructor is private.
</details>
Ah, true, I overlooked that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905)*
<!-- Sent from Reviewable.io -->

mimimi085181 commented:
String nitpicking:
1. "optionally" collects data. That sounds a bit weird. I think the text itself should reflect the fact that's it's voluntary, and the word optional there just sounds weird. But maybe it's just me...
2. It's not clear whether the game list or the played games are send. From what i gather, it sends which games are played, right? So that should be listed. And if the game list is not sent, and that's by decision, it should be mentioned as well
3. Right now the text sounds like it sends data once, which isn't true(?).
4. "no private data". If the data is sufficient to identify a user, you can argue that it counts as private data. Right now, people just have to trust this isn't the case. I think it should be made pretty clear what data is send and what isn't.
5. "Generate a New Statistics Identity". Maybe new in lower case? I know we are using a weird system for upper and lower case, well weird for english... But new in upper case seems wrong.

Oh, and i want to see the unique string in the gui, copyable, so i can check that i actullay got a new one. Also, it's weird to get one, if you are not agreeing to sending the data. Maybe that should be greyed out for people who don't opt-in?

(not reviewing the 1st tooltip in the config:)
How about string recycling and putting the 1st dialog behind a button in the config? Right now, there's 2 sets of strings exlaining the same thing. The unique id generation could be put into the same dialog, and actually allowing the data to be sent should only be possible when an id was generated. So the user has to do 2 things to actually allow this, which makes it even clearer that this is voluntary opt-in.

Maybe actually use the phrase "opt-in"?

Last but not least, i'd like to be able to review the data that is send, before it actually is sent. But that's really complicated and i understand why it won't be done before 5.0.

JosJuice commented:
> 5. "Generate a New Statistics Identity". Maybe new in lower case? I know we are using a weird system for upper and lower case, well weird for english... But new in upper case seems wrong.

No, new in uppercase sounds right. Adjectives are in uppercase.

delroth commented:
> 1. "optionally" collects data. That sounds a bit weird. I think the text itself should reflect the fact that's it's voluntary, and the word optional there just sounds weird. But maybe it's just me...

It was changed to optionally because of another comment complaining about the opposite thing. I think we can nitpick the wording for weeks and we still won't agree.

> 2. It's not clear whether the game list or the played games are send. From what i gather, it sends which games are played, right? So that should be listed. And if the game list is not sent, and that's by decision, it should be mentioned as well

We can't list everything in a tiny message box. Ideally we'd have a link with more documentation, but WX sucks and doesn't allow us to do that. People that are uncomfortable with the amount of info we give can press "No".

> 3. Right now the text sounds like it sends data once, which isn't true(?).

What makes you think that? It explicitly says "while running the emulator".

> 4. "no private data". If the data is sufficient to identify a user, you can argue that it counts as private data. Right now, people just have to trust this isn't the case. I think it should be made pretty clear what data is send and what isn't.

That's not how most people define PII. Also see response to 2.

> Last but not least, i'd like to be able to review the data that is send, before it actually is sent. But that's really complicated and i understand why it won't be done before 5.0.

Never going to happen. From a UX standpoint it's terrible, and we plan to report e.g. performance data in the background while a game is being played. Frankly, if you care so much that having access to the source code is not enough, and you want to explicitly check what is being sent, press "no". Getting data from the low % of users that want this amount of control is not worth spending so much time on the implementation.

mbc07 commented:
Would it be possible to add banner uploading in a follow-up PR (after 5.0, of course)? Should address [issue 8283] (https://bugs.dolphin-emu.org/issues/8283) and help gathering missing banners for the compatibility page on the main website...

phire commented:
On Jun 19, 2016 6:02 AM, "Pierre Bourdon" <notifications@github.com> wrote:
>>
>> "optionally" collects data. That sounds a bit weird. I think the text
itself should reflect the fact that's it's voluntary, and the word optional
there just sounds weird. But maybe it's just me...
>
> It was changed to optionally because of another comment complaining about
the opposite thing. I think we can nitpick the wording for weeks and we
still won't agree.

Before it said "occasionally" which is not really the opposite of
"optionally" and had much worse issues.

If there is any issue which could have benefited from months of nitpicking,
it's this one.

>>
>> It's not clear whether the game list or the played games are send. From
what i gather, it sends which games are played, right? So that should be
listed. And if the game list is not sent, and that's by decision, it should
be mentioned as well
>
> We can't list everything in a tiny message box. Ideally we'd have a link
with more documentation, but WX sucks and doesn't allow us to do that.
People that are uncomfortable with the amount of info we give can press
"No".
>>
>> Right now the text sounds like it sends data once, which isn't true(?).
>
> What makes you think that? It explicitly says "while running the
emulator".
>>
>> "no private data". If the data is sufficient to identify a user, you can
argue that it counts as private data. Right now, people just have to trust
this isn't the case. I think it should be made pretty clear what data is
send and what isn't.
>
> That's not how most people define PII. Also see response to 2.
>>
>> Last but not least, i'd like to be able to review the data that is send,
before it actually is sent. But that's really complicated and i understand
why it won't be done before 5.0.
>
> Never going to happen. From a UX standpoint it's terrible, and we plan to
report e.g. performance data in the background while a game is being
played. Frankly, if you care so much that having access to the source code
is not enough, and you want to explicitly check what is being sent, press
"no". Getting data from the low % of users that want this amount of control
is not worth spending so much time on the implementation.
>
> —
> You are receiving this because you commented.
>
> Reply to this email directly, view it on GitHub, or mute the thread.

phire commented:
> >  "optionally" collects data. That sounds a bit weird. I think the text itself should reflect the fact that's it's voluntary, and the word optional there just sounds weird. But maybe it's just me...

> It was changed to optionally because of another comment complaining about the opposite thing. I think we can nitpick the wording for weeks and we still won't agree.

Before it said "occasionally" which is not really the opposite of "optionally" and had much worse issues.

If there is any issue which could have benefited from months of nitpicking, it's this one.

*Reposted because github did not like the email formatting on this.*

delroth commented:
From the code side I think this PR is ready. I'll let the decision about the wording to a native speaker. We should get this merged ASAP in order to have the strings ready for translation early.

phire commented on Source/Core/DolphinWX/Config/GeneralConfigPane.cpp:
```diff
@@ -55,19 +58,25 @@ void GeneralConfigPane::InitializeGUI()
 	m_cheats_checkbox->SetToolTip(_("Enables the use of Action Replay and Gecko cheats."));
 	m_force_ntscj_checkbox->SetToolTip(_("Forces NTSC-J mode for using the Japanese ROM font.\nIf left unchecked, Dolphin defaults to NTSC-U and automatically enables this setting when playing Japanese games."));
+	m_analytics_checkbox->SetToolTip(_("Enables the collection and sharing of usage statistics data with the Dolphin development team. This data is used to improve the emulator and help us understand how our users interact with the system. No private data is ever collected."));
+	m_analytics_new_id->SetToolTip(_("Usage statistics reporting uses a unique per-machine identifier to distinguish users from one another. This button generates a new identifier for this machine which is dissociated from the previous one."));
```
Perhaps want to mention the identifier is randomly generated and contains no identifying information, like CPU or HDD serial numbers.

phire commented:
Here is a screenshot of the current dialog on windows:

![image](https://cloud.githubusercontent.com/assets/138484/16173880/38877428-3603-11e6-8a26-63842c86c603.png)

phire commented:
It might make more sense for the options to go on the interface tab of configuration, rather than General (currently it's making general tap very tall)

As for the dialog, I think the question should go last. perhaps rephrase it as:

    Dolphin optionally collects data about the hardware, operating system,
    feature usage, and application configuration while running the emulator.

    No private data is ever collected. This data helps us understand how
    people use Dolphin and prioritize our efforts. Any authorization can be
    revoked later in the Dolphin Settings.

    Do you authorize dolphin to report this information to Dolphin developers?
                                                                  Yes |  No

Other possible changes that I'm less sure on:

`how people use Dolphin` to `how people and emulated games use Dolphin`

`feature usage, and application configuration` to `application configuration and hardware features used by emulated games`

Helios747 commented:
ping

jimbo1qaz commented:
not sure if this comment is relevant...
Personally I would be more likely to accept data gathering, to help identify specific configurations that lead to bugs or instability, or advising users about non-recommended settings.

phire commented:
That's it for my User Interface nitpicks.

But I have major concerns that this is collecting data which it simply doesn't need to collect, which could be used for de-anonymizing attacks. There is no reason to collect all games a single user plays and the times that they are playing them under a single unique identifier. 

Instead, we should generate a unique 128bit (or larger) identifier and keep it private. Then we can hash the game ID to generate a unique per-game ID.

Startup events will be hashed with a constant value, so we can still get accurate usage counts. This is a simple solution that stops us sending data to the server that we didn't want to collect in the first place and we don't lose any of the data we do want to collect.

We also should ask the question of if we need to track users across their entire lifetime. Hashing the Week (or month) into the reported user IDs would still allow us to measure the instantaneous user count (X users used this version of dolphin on Windows Vista this month) while preventing entire lifetime tracking.

Do we want to track users across multiple versions of dolphin? The user's entire path through a bisect will be reported under the same ID, which is a uniquely identifiable piece of information (connected with their bug report). A similar scheme could be used here.

Do we really need to report every single boot? We can record the last time a boot report was sent and only send one once a week.

I know you have plans to make sure the backend will only accept queries that don't identify users. But its not like people can audit our backend and servers. But people can audit our code, and I'd rather modify things so infomation we don't want is never sent.

This is about protecting our users privacy, and avoiding any bad press to maximise the amount of non-anonymous data we collect.

Nebulous112 commented:
How about change the first paragraph to:

"If authorized, when running the emulator Dolphin can collect data on its performance, feature usage, and application configuration, as well as data on your system's hardware and operating system."

The other paragraphs seem fine.

As a native English speaker, I have to agree that the original "optionally" sounds unwieldy. The original placement of "application configuration" could be misinterpreted to mean data on all of the user's programs.

JMC47 commented:
I agree with @Nebulous112's suggestions.

phire commented:
Updated version with @Nebulous112 and @jimbo1qaz's suggestions:

    If authorized, Dolphin can collect data on its performance, feature
    usage, and configuration, as well as data on your system's hardware
    and operating system.

    No private data is ever collected. This data helps us understand how
    people and emulated games use Dolphin and prioritize our efforts.
    It also helps us identify rare configurations that are causing bugs,
    performance and stability issues. This authorization can be revoked
    at any time through Dolphin's settings.

    Do you authorize dolphin to report this information to Dolphin's developers?
                                                                  Yes |  No

JMC47 commented:
lgtm

Nebulous112 commented:
Sorry, one more idea:

```
If authorized, Dolphin can collect data on its performance, feature
usage, and configuration, as well as data on your system's hardware
and operating system.

No private data is ever collected. This data helps us understand how
people and emulated games use Dolphin and allows us to prioritize our efforts.
It also helps us identify rare configurations that are causing bugs,
performance and stability issues.

This authorization can be revoked at any time through Dolphin's settings.

Do you authorize dolphin to report this information to Dolphin's developers?
                                                              Yes |  No
```

"prioritize our efforts" --> "allows us to prioritize our efforts"

Line break for "This authorization can be revoked..." makes it stand out more.

"Do you authorize dolphin..." --> "Do you authorize Dolphin..." (capitalization)

mbc07 commented:
> If authorized, Dolphin can collect data on its performance, feature
usage, and configuration, as well as data on your system's hardware
and operating system.

AFAIK there's a comma here that is wrong (between feature usage and configuration). In other words, 

"If authorized, Dolphin can collect data on its performance, **feature
usage and configuration**, as well as data on your system's hardware
and operating system."

phire commented:
> "prioritize our efforts" --> "allows us to prioritize our efforts"

It's just extra words that don't add anything and make the dialog box longer. But the other fixes are good.

phire commented:
@mbc07 everyone is doing the oxford comma these days.

jimbo1qaz commented:
...Collect data on Dolphin's (performance, feature usage, and
configuration)?

If so, keep the comma.

On Sat, Jun 18, 2016 at 4:55 PM, Mateus B. Cassiano <
notifications@github.com> wrote:

> If authorized, Dolphin can collect data on its performance, feature
> usage, and configuration, as well as data on your system's hardware
> and operating system.
>
> AFAIK there's a comma here that is wrong (between feature usage and
> configuration). In other words,
>
> "If authorized, Dolphin can collect data on its performance,
> *feature usage and configuration*, as well as data on your system's
> hardware
> and operating system."
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dolphin-emu/dolphin/pull/3905#issuecomment-226971449>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA3yJavnwiZfovW4TA6AQL2Bukh5D5Wfks5qNIVagaJpZM4I41vv>
> .
>

phire commented:
Do we want to use a semicolon? 

    If authorized, Dolphin can collect data on its performance, feature
    usage, and configuration; as well as data on your system's hardware
    and operating system.

(I'm not really an expert on semicolon usage)

jimbo1qaz commented:
semicolon? probably not.

this is more suited for chat than comments? (does reviewable show comments as a chat window?)

Is there a comprehensive list of what data is being collected?

Nebulous112 commented:
Agree to keep the comma. And no, a broad rule for semicolons are that they should only be used if both parts could also be used as individual sentences. It is a bit of a run-on sentence though.

mbc07 commented:
> @mbc07 everyone is doing the oxford comma these days.

Didn't know that, had to Google it. No worries then, keep the comma...

jimbo1qaz commented:
> Do we really need to report every single boot? We can record the last time a boot report was sent and only send one once a week.
> 
> I know you have plans to make sure the backend will only accept queries that don't identify users. But its not like people can audit our backend and servers. But people can audit our code, and I'd rather modify things so infomation we don't want is never sent.

Seconded. I really don't want Dolphin to know when I'm running games, or not.

Sonicadvance1 commented:
Pro tip, skew the stats by generating scripts to start multiple copies of dolphin and restart after the systems amount of time before it sends data. 
Abuse it to the maximum

JMC47 commented:
It doesn't know you're running games, it's just recording that someone booted a game.

phire commented on Source/Core/Core/Analytics.cpp:
```diff
@@ -0,0 +1,237 @@
+	Common::AnalyticsReportBuilder builder(m_base_builder);
+	builder.AddData("type", "dolphin-start");
+	builder.AddData("ui-type", ui_type);
+	builder.AddData("id", MakeUniqueId("game-start"));
```
Tiny nitpick, this should be "dolphin-start" not "game-start".

Doesn't affect implementation just self-documentation.

delroth replied:
Fixed.

phire commented:
Privacy Update:

The latest version of this PR generates a random 128bit secret on the first boot. User IDs are now generated based on a SHA1 hash of the secret and the game ID, so each game will look like it's being launched by a different user. It won't be possible to tie a user to a list of played games anymore.

The boot reports all use a single ID, so we still get the count of unique dolphin users.

phire commented:
Reviewed 301 of 301 files at r6.
Review status: 2 of 313 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3905)*
<!-- Sent from Reviewable.io -->

CarlKenner commented:
What happens with this analytics code in other branches of Dolphin (eg. Dolphin VR)?

JMC47 commented:
You can change the distributor so it shows up as Dolphin VR.

phire commented:
The other branches will have unique hashes too, so we can identify them even if they don't change the distributor.

mbc07 commented:
Any ETA on when the new strings will be available in Transifex?

delroth commented:
@mbc07 how about now? :)

CarlKenner commented:
Now we have a file called share.h in the include directories, and xiosbase.h (somehow included from SFML_Network) is _not happy_ about it, at least in VS2013. Share.h is a standard include file that should not be overridden. xiosbase.h uses triangle brackets to include it, while curl uses quotation marks, so I don't know if there's a way to get it to look in different folders when using triangle brackets rather than quotes.

lioncash commented:
VS 2015 is listed as a prerequisite in the readme and building instructions. VS 2013 is not readily supported (and will likely be eventually completely incompatible when the next VS version supporting C++17 library additions comes out).

Fallcrest commented:
There are a lot of problems with this that will have to be ironed out when considering future versions of this. How will we handle collecting information about Netplay connections?

JMC47 commented:
Right now all we collect is the type of session.  Movie or Netplay are specified.

Fallcrest commented:
@JMC47 Okay. But there are a lot of variables at play when attempting to figure out why something in Netplay failed. Will information from two people connected in Netplay be linked together somehow? What would happen if one person opted into the information collecting and one person did not?

CarlKenner commented:
I fixed the share.h issue. I know you don't support VS2013, but you still need to fix it because it will probably cause issues for you down the line with other projects (if it isn't already).

This line needs to be moved from VSProps/Base.props to Externals/curl/curl.vcxproj:
      <AdditionalIncludeDirectories>$(ExternalsDir)curl\lib;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>

The lib folder is not intended to be used by other projects, it is only for building curl itself. The include folder is for other projects.

Dolphin VR is set up to be able to be built with either VS2013 or VS2015. It can no longer save videos though if built in VS2013 because ffmpeg is not available in source code form that people can build, and you only include a VS2015 library. Official releases will use VS2015, but I want to keep VS2013 support in case people need to use VS2013 libraries such as older versions of some VR libraries.

VS2013 is already completely incompatible due to us using constexpr everywhere. But in Dolphin VR I have #ifdef around each one with a replacement for VS2013. Also there's no DX12 support in VS2013.

CarlKenner commented:
For me, Dolphin never terminates, because it gets stuck waiting forever on this line of AnalyticsReporter::~AnalyticsReporter():

m_reporter_thread.join();

phire commented:
Have you done anything weird to Common::Event or Common::Flag in your vs2013 modifications?

Breaking either of those could cause AnalyticsReported::ThreadProc() to not exit when requested.
